File: shared/schema.py
"""Backend ↔ Worker 간 공유되는 Pydantic 스키마.

이 모듈의 모델은 두 컨테이너(backend, worker) 모두에서 import 된다.
변경 시 frontend의 src/types/schema.ts 도 함께 갱신해야 한다.

worker-origin 노트북이 사용하는 한글 컬럼('모선항차', '접안위치(F)' 등)을
영문 식별자로 매핑해 노출한다 — 한글 컬럼은 worker 내부 DataFrame 변환 시점에서만 쓴다.
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


JobStatus = Literal["pending", "running", "succeeded", "failed"]
SolverName = Literal["gurobi", "cqm", "hybrid"]

# BPTC 선석배정 그래픽(G) 페이지 VslMsg(...) 의 8번째 인자 plan_cd 매핑.
# 사이트 JS 안에 정의된 한국어 라벨:
#   L → "적하 프래닝까지 완료"
#   D → "양하 프래닝까지 완료"
#   C → "크래인배정 완료"
#   (빈값/기타) → "크래인미 배정"
PlanStatus = Literal[
    "loading_planned",      # plan_cd=L
    "discharge_planned",    # plan_cd=D
    "crane_assigned",       # plan_cd=C
    "crane_unassigned",     # 그 외
]


class BPTRecord(BaseModel):
    """Berth Productivity Table 한 행 — 노트북의 BPT_Result.xlsx 한 행과 동치."""

    vessel_id: str = Field(..., description="모선항차")
    length: int = Field(..., gt=0, description="선박 길이(m)")
    eta_int: float = Field(..., description="ETA_int — 도착 예정 시각(시간)")
    etb_int: float = Field(..., description="ETB_int — 현재 배치 접안 시각")
    etd_int: float = Field(..., description="ETD_int — 현재 배치 출항 시각")
    berth_position: float = Field(..., description="접안위치(F) — 선석 시작 좌표(m)")
    yangha_van: float = Field(0, description="양하(Van)")
    seonjeok_van: float = Field(0, description="선적(Van)")
    plan_status: PlanStatus | None = Field(
        None,
        description=(
            "선석배정 그래픽 페이지의 plan_cd 매핑 — "
            "loading_planned(적하 프래닝 완료) / discharge_planned(양하 프래닝 완료) / "
            "crane_assigned(크래인배정 완료) / crane_unassigned(크래인미 배정). "
            "BP 그래픽에 해당 선박이 없으면 None."
        ),
    )


class ScheduleEntry(BaseModel):
    """최적화 결과 한 행."""

    vessel_id: str
    length: int
    eta: float
    etb: float
    etd: float
    berth_position: float
    note: str = ""


class OptimizeRequest(BaseModel):
    """최적화 요청 (backend → worker)."""

    job_id: str = ""
    bpt_records: list[BPTRecord]
    solver: SolverName = "gurobi"
    planning_start_time: float = Field(
        0, description="시뮬레이션 기준 시점(시간). 노트북 기본값 14.5 또는 80."
    )


class OptimizeResult(BaseModel):
    """최적화 결과/진행 상태 (worker → backend → frontend).

    status가 running 일 때 schedule 은 비어 있고 started_at 으로 경과 시간을
    계산할 수 있다. status가 succeeded 일 때 elapsed_seconds 가 채워진다.
    """

    job_id: str
    status: JobStatus
    schedule: list[ScheduleEntry] = []
    objective_value: float | None = None
    elapsed_seconds: float | None = None
    started_at: datetime | None = None
    error_message: str | None = None


class JobAccepted(BaseModel):
    """POST /jobs/ 의 즉시 반환 응답 (202 Accepted)."""

    job_id: str
    status: JobStatus = "running"

File: frontend/src/shared/types/schema.ts
// shared/schema.py (Pydantic) 와 동기화 필요 — 변경 시 양쪽 모두 갱신.
// 추후 openapi-typescript 로 자동 생성 도입 예정 (AGENTS.md §4.1).
//
// zod 스키마는 런타임 응답 검증용. 타입은 z.infer 로 파생한다.

import { z } from 'zod';

export const SOLVER_NAMES = ['gurobi', 'cqm', 'hybrid'] as const;
export type SolverName = (typeof SOLVER_NAMES)[number];

export const JOB_STATUSES = ['pending', 'running', 'succeeded', 'failed'] as const;
export type JobStatus = (typeof JOB_STATUSES)[number];

export const BPTRecordSchema = z.object({
  vessel_id: z.string().min(1),
  length: z.number().int().positive(),
  eta_int: z.number(),
  etb_int: z.number(),
  etd_int: z.number(),
  berth_position: z.number(),
  yangha_van: z.number().default(0),
  seonjeok_van: z.number().default(0),
});
export type BPTRecord = z.infer<typeof BPTRecordSchema>;

export const ScheduleEntrySchema = z.object({
  vessel_id: z.string(),
  length: z.number(),
  eta: z.number(),
  etb: z.number(),
  etd: z.number(),
  berth_position: z.number(),
  note: z.string().default(''),
});
export type ScheduleEntry = z.infer<typeof ScheduleEntrySchema>;

export const OptimizeRequestSchema = z.object({
  job_id: z.string().optional(),
  bpt_records: z.array(BPTRecordSchema),
  solver: z.enum(SOLVER_NAMES),
  planning_start_time: z.number().default(0),
});
export type OptimizeRequest = z.infer<typeof OptimizeRequestSchema>;

export const OptimizeResultSchema = z.object({
  job_id: z.string(),
  status: z.enum(JOB_STATUSES),
  schedule: z.array(ScheduleEntrySchema).default([]),
  objective_value: z.number().nullable().default(null),
  elapsed_seconds: z.number().nullable().default(null),
  started_at: z.string().nullable().default(null),
  error_message: z.string().nullable().default(null),
});
export type OptimizeResult = z.infer<typeof OptimizeResultSchema>;

export const JobAcceptedSchema = z.object({
  job_id: z.string(),
  status: z.enum(JOB_STATUSES).default('running'),
});
export type JobAccepted = z.infer<typeof JobAcceptedSchema>;

// GET /jobs/ — 백엔드가 list[dict] 로 내려주므로 손으로 정의한 메타.
export const JobMetaSchema = z.object({
  job_id: z.string(),
  status: z.enum(JOB_STATUSES),
  solver: z.enum(SOLVER_NAMES),
  objective_value: z.number().nullable(),
  elapsed_seconds: z.number().nullable(),
  created_at: z.string().nullable(),
  started_at: z.string().nullable(),
  completed_at: z.string().nullable(),
});
export type JobMeta = z.infer<typeof JobMetaSchema>;
